pkg/reconciler/taskrun/taskrun_test.go

```diff
@@ -4657,3 +4672,96 @@ func objectMeta(name, ns string) metav1.ObjectMeta {
 		Annotations: map[string]string{},
 	}
 }
+
+func TestReconcile_validateTaskRunResults(t *testing.T) {
+	taskRunResultsTypeMismatched := parse.MustParseTaskRun(t, `
+metadata:
+  name: test-taskrun-results-type-mismatched
+  namespace: foo
+spec:
+  taskRef:
+    name: test-results-task
+status:
+  taskResults:
+    - name: aResult
+      type: string
+      value: aResultValue
+`)
+
+	taskRunResultsObjectMissKey := parse.MustParseTaskRun(t, `
+metadata:
+  name: test-taskrun-results-object-miss-key
+  namespace: foo
+spec:
+  taskRef:
+    name: test-results-task
+status:
+  taskResults:
+    - name: aResult
+      type: array
+      value:
+       - 1
+       - 2
+    - name: objectResult
+      type: object
+      value:
+        url: abc
+`)
+
+	taskruns := []*v1beta1.TaskRun{
+		taskRunResultsTypeMismatched, taskRunResultsObjectMissKey,
+	}
+
+	d := test.Data{
+		TaskRuns: taskruns,
+		Tasks:    []*v1beta1.Task{resultsTask},
+		ConfigMaps: []*corev1.ConfigMap{{
+			ObjectMeta: metav1.ObjectMeta{Namespace: system.Namespace(), Name: config.GetFeatureFlagsConfigName()},
+			Data: map[string]string{
+				"enable-api-fields": config.AlphaAPIFields,
+			},
+		}},
+	}
+	for _, tc := range []struct {
+		name             string
+		taskRun          *v1beta1.TaskRun
+		wantFailedReason string
+		expectedError    error
+	}{{
+		name:             "taskrun results type mismatched",
+		taskRun:          taskRunResultsTypeMismatched,
+		wantFailedReason: podconvert.ReasonFailedValidation,
+		expectedError:    fmt.Errorf("1 error occurred:\n\t* missmatched Types for these results, map[aResult:[array]]\n\n"),
```

Comment:
This is a confusing error message because it's not clear which result (taskrun result or result declared in task spec) is an array and which is not, and what the type of the result that is not an array is. Can it be improved?

Reply:
The current code logic will only collect the wrong results/params names and log them, and same for other parts. You're 100% correct that this can be improved! I will think about it and maybe address it in a new PR? If that is ok I will open an issue to track

https://github.com/tektoncd/pipeline/blob/8a7b0cfa755038f4cfdcd88c314a72a90bcab1a2/pkg/reconciler/pipelinerun/resources/validate_params.go#L82-L90

Reply:
Sure, you can definitely change in a separate PR-- it's a small enough change I'll leave it up to you whether to create an issue or just open a follow-up PR